### ISSUE TITLE
How to use SfMPerspectiveCameras for renderer?

### ISSUE THREAD

**Issue by wangg12** [facebookresearch/pytorch3d#18]
I want to use this differentiable renderer to render a [LINEMOD object](http://ptak.felk.cvut.cz/6DB/public/bop_datasets/lm_models.zip). The .ply model vertices are loaded and then scale from `mm` to `m` by a factor 0.001.

The camera intrinsic: 
```
K = np.array([[572.4114, 0, 325.2611], [0, 573.57043, 242.04899], [0, 0, 1]])
```

Rotation and translation
```
R = np.eye(3)
T = np.array([-0.1, 0.1, 0.7])
```

I followed the camera position tutorial to set the cameras and renderers, except that the camera I used is
```
cameras = SfMPerspectiveCameras(
        focal_length=((K[0,0], K[1,1]), ),
        principal_point=((K[0,2], K[1,2]),),
        device=device,
    )
```
And since I want to get 480x640 image, I set the `image_size=640` and crop to `480x640` after rendering.

However, there is nothing in the rendered results. I could get correct results with a OpenGL renderer with the same settings.

So I wonder how I can correctly use the `SfMPerspectiveCameras`?

**Comment by davnov134** [comment]
Hi wangg12,

can I please ask whether you can share the rendering script?

More specifically, it is not obvious what is the position of the LM object in the world coordinates. Furthermore, the "R" matrix you are using does not seem to be a valid rotation, see below:
```
R = torch.tensor([[-0.5,  0.8,  0],  [0, 0, -1],  [-0.8, -0.5,  0]])
torch.det(R)
> tensor(0.8900) # instead of 1
```

In case the LM object should be centered around (0, 0, 0), I would suggest setting the camera to a trivial one with a non-zero z-component of the `T` vector, e.g.:
```
R = torch.eye(3), T = torch.tensor([0., 0., 2.])
```
This would point the camera directly at the object center and render it from the distance of 2 world units.

**Comment by wangg12** [comment]
Sorry, `R` was not the correct one. It was transformed from axis angle and should be correct. I shared my script [here](<script src="https://gist.github.com/wangg12/32250e352671c7f8c9a42ddb437142a6.js"></script>)

**Comment by wangg12** [comment]
@nikhilaravi It seems if I devide the `fx,fy,px,py` in `K` by image size (640), there is some wrong rendered result on the image.
(left: p3d renderer; right: opengl renderer)
![image](https://user-images.githubusercontent.com/9958446/73381203-9be9c200-42c5-11ea-9313-665406de7f30.png)

**Comment by wangg12** [comment]
Finally I got the correct results with a modified OpenGL Real Perspective Camera.
The scripts are [here](https://gist.github.com/wangg12/32250e352671c7f8c9a42ddb437142a6).
![image](https://user-images.githubusercontent.com/9958446/73388548-bbd3b280-42d2-11ea-8ed7-ab9714169757.png)

BTW, @nikhilaravi How can I render the differentiable depth values?

**Comment by aluo-x** [comment]
Haven't tried this myself since I am working with RGB, but [Line 37 of renderer.py](https://github.com/facebookresearch/pytorch3d/blob/dbf06b504b525c7f6680c5709b63df6413616d2e/pytorch3d/renderer/mesh/renderer.py#L37) exposes a fragment, which should have a [zbuf object](https://github.com/facebookresearch/pytorch3d/blob/dbf06b504b525c7f6680c5709b63df6413616d2e/pytorch3d/renderer/blending.py#L110).

**Comment by nikhilaravi** [comment]
@wangg12 Glad you got it to work. I can take a look at your script later! For future, please try to share the minimum amount of code needed to explain the issue so we can review more quickly :) 

Thanks @aluo-x for explaining the zbuf/fragments :)  As mentioned, we expose the zbuffer as part of the `fragments` object which is output from the rasterizer.  It will be a `N x H x W x K` dimensional tensor.

**Comment by bhadresh74** [comment]
@wangg12 Thanks for sharing your script. I am trying to do something similar using your script. 
However the rendering is vertically stretched as shown below.
<img width="340" alt="Screen Shot 2020-03-24 at 5 35 07 PM" src="https://user-images.githubusercontent.com/9024246/77483365-fe5acc80-6df5-11ea-9478-0aef8ef92ca7.png">
Ideally, it should look like this.
<img width="337" alt="Screen Shot 2020-03-24 at 5 34 56 PM" src="https://user-images.githubusercontent.com/9024246/77483443-25190300-6df6-11ea-8479-dccec9a5fe25.png">

My image size is (960, 1280). I believe it has something to do with rasterizer since it only works for square image. 
Can you guide me in the right direction here?

**Comment by nikhilaravi** [comment]
@bhadresh74 yes the rasterizer currently only works for square images!

**Comment by bhadresh74** [comment]
@nikhilaravi Thanks for the prompt response. Do you think the rendering is stretched due to rasterization? If not, then I can keep debugging to resolve this.
